### PR TITLE
Include information from GHCR response bodies in error reporting

### PR DESCRIPTION
**Pull request by conorsloan**
Previously, when GHCR would return an unexpected status code (5xx, 4xx) our logging would say something like:

> Unexpected response from PUT manifest 400

Which doesn't tell the user much (in this case, it is useful for the user to know that the tag has already been used!).

This change introduces parsing of the GHCR response body format for errors, which are then included in the error we log in the actions run. The error message now looks like:

> Unexpected 400 Bad Request response from manifest upload. Errors: BAD_REQUEST - A container with version 2.0.5 has previously existed as an Actions Package and cannot be reused.

This should improve a user's ability to understand why a run failed and what their next steps are.

### CODE REVIEW

File: src/ghcr-client.ts [hunk 1]
```diff
@@ -200,6 +206,57 @@ async function uploadManifest(
...
+  } catch (error) {
+    // Ignore error
+  }
+
+  return `Unexpected ${response.status} ${response.statusText} response from ${requestDescription}. ${errorString}`
+}
+
+// Runtime checks that parsed JSON object is in the expected format
+// {"code": "BAD_REQUEST", "message": "Something went wrong."}
+function isGHCRError(obj: unknown): boolean {
```
Comment (conorsloan) [inline src/ghcr-client.ts:249]: Reviewers 👋 please let me know if there's a better way to do runtime checks than this. I am not a javascript developer 😄

File: src/ghcr-client.ts [hunk 2]
```diff
@@ -200,6 +206,57 @@ async function uploadManifest(
   return digestResponseHeader
 }
 
+interface ghcrError {
+  code: string
+  message: string
+}
+
+// Generate an error message for a failed HTTP request
+async function errorMessageForFailedRequest(
```
Comment (Jcambass) [inline src/ghcr-client.ts:215]: Does this really need to be `async`?

File: src/ghcr-client.ts [hunk 3]
```diff
@@ -200,6 +206,57 @@ async function uploadManifest(
...
+    ) {
+      const errorMessages = errors.map((error: ghcrError) => {
+        return `${error.code} - ${error.message}`
+      })
+      errorString = `Errors: ${errorMessages.join(', ')}`
+    }
+  } catch (error) {
+    // Ignore error
+  }
+
```
Comment (Jcambass) [inline src/ghcr-client.ts:243]: I would have expected a simpler form where we just part the JSON and then try to navigate to the property we care about using the [optional chaining operators](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining).

File: src/ghcr-client.ts (hunk 2 above)
Comment (conorsloan) [reply src/ghcr-client.ts:215]: Yeah because of `const bodyText = await response.text()`

File: src/ghcr-client.ts (hunk 3 above)
Comment (conorsloan) [reply src/ghcr-client.ts:243]: The issue here is that this is a list of errors, some of which theoretically might not contain the appropriate schema. If the errors don't all have the right fields, then we want to just fall all the way back to printing the whole response body, or some information could easily be lost.

The approach I've gone with is either it's all in the correct format or we should just use the raw response. 
I'm not sure how optional operators would help in that case.

**Pull request merged**